packages/core/src/util/query-result.js

```diff
@@ -15,15 +22,41 @@ export class QueryResult extends Promise {
     });
     this._resolve = resolve;
     this._reject = reject;
+    this._state = QueryState.pending;
+    this._value = undefined;
   }
 
   /**
-   * Resolve the result Promise with the provided value.
+   * Resolve the result Promise with a prepared value or the provided value.
+   * This method will only succeed if either a value is provided or prepared.
    * @param {*} value The result value.
    * @returns {this}
    */
   fulfill(value) {
-    this._resolve(value);
+    if (this._value !== undefined) {
+      if (value !== undefined) {
+        throw Error('Promise has prepared and provided value');
+      }
+      this._resolve(this._value);
+    } else if (value === undefined) {
+      throw Error('Promise has has neither prepared nor provided value');
+    } else {
+      this._resolve(value);
+    }
+
+    this._state = QueryState.done;
+
+    return this;
+  }
+
+  /**
+   * Prepare to resolve the result Promise with the provided value.
+   * @param {*} value The result value.
+   * @returns {this}
+   */
+  prepare(value) {
+    this._state = QueryState.prepared;
```

Comment:
`QueryState.ready` might be a reasonable alternative to `prepared`?

Reply:
Sounds good. Should the method still be called `prepare` or become `ready`?

Reply:
I'd rename it to `ready`

Reply:
done